### PR TITLE
Reaction lookup and Addiction refactoring [PORT from TG]

### PR DESCRIPTION
**Pull request by dwasint**
# About The Pull Request
Ports the reaction lookup window from tg https://github.com/tgstation/tgstation/pull/56868
Also ports TG's Addiction rework  https://github.com/tgstation/tgstation/pull/56923

For developers: addiction stages work very similar to disease stages if you want to add a cool effect you can look at disease code aswell

## Why It's Good For The Game
Makes chemistry easier for newer players who don't have a wiki open at all times

Also the current addiction mechanic has many serious flaws. First of all it is easy to cheat. The threshhold mechanic allows you take huge amounts of drugs and yet avoid addiction by structuring your doses.

This makes addiction only a concern of clueless newbies and those who have drugs forcibly injected into them, not drug-fueled powergamers.

In addition, addiction is hyperspecific and often quite lethal. This makes seeking out medical treatment in order to kick the habit almost invariably the best option.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

</details>

## Changelog

:cl:
add: All drug addictions are now put into 5 categories: Stimulants, Opiods, Alcohol, hallucinogens, and maintenance drugs. There is no longer an addiction threshold and becoming addicted is done by gaining addiction points from doing drugs in a specific category, if you become addicted you can stop being addicted by lowering your addiction points until it reaches a low enough value.
add: Adds a new reaction lookup window to the chem dispenser
/:cl:

### CODE REVIEW

File: code/modules/reagents/chemistry/reagents/alcohol_reagents.dm [hunk 1]
```diff
@@ -35,6 +35,10 @@ All effects don't start immediately, but rather get worse over time; the rate is
 91-100: Dangerously toxic - swift death
 */
 
+/datum/reagent/consumable/ethanol/New()
+	addiction_types = list(/datum/addiction/alcohol = 0.05 * boozepwr)
```
Comment (nednaZ) [inline code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:39]: Would this mean that the Crevice Spike and Fanciulli actually reduce your addiction to alcohol every tick?

Also, that puts this at a range of -0.5 to 15 a tick, if you want to add a comment here.

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 2]
```diff
@@ -1522,6 +1525,7 @@
 	taste_mult = 1.5
 	process_flags = ORGANIC | SYNTHETIC
 	random_unrestricted = FALSE
+	addiction_types = null
```
Comment (nednaZ) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:1528]: Does this need a null addiction type?

File: code/modules/reagents/withdrawl/_addiction.dm [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		if(1)
+			withdrawal_stage_1_process(affected_carbon, delta_time)
+		if(2)
+			withdrawal_stage_2_process(affected_carbon, delta_time)
+		if(3)
+			withdrawal_stage_3_process(affected_carbon, delta_time)
+
+	LAZYADDASSOC(affected_carbon.mind.active_addictions, type, 1 * delta_time) //Next cycle!
+
+/// Called when addiction enters stage 1
```
Comment (nednaZ) [inline code/modules/reagents/withdrawl/_addiction.dm:107]: These stages are very similar to disease stages.
Could probably point this out in the changelog for future coders, can put interesting effects with new chems.

**Review by nednaZ — COMMENTED** [review]
Code itself is reviewed.
I'll need to test it myself and run a test merge in the future however.

File: code/__HELPERS/reagents.dm [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+		return get_chem_id(input)
+
+/proc/find_reagent_object_from_type(input)
+	if(GLOB.chemical_reagents_list[input]) //prefer IDs!
+		return GLOB.chemical_reagents_list[input]
+	else
+		return null
+
+///Returns a random reagent object minus blacklisted reagents
+/proc/get_random_reagent_id()
```
[the remark this answers is not in the document]
Comment (dwasint) [reply code/__HELPERS/reagents.dm:29]: this was just moved from where we had it before and shrunk down, but really this probably shouldn't be shrunk at all but it is used in vent clogs egg snacks and seeds

**Review by nednaZ — COMMENTED** [review]
Test merging

**Review by nednaZ — APPROVED** [review]
No unfixed issues found during testmerge

**Pull request merged**